### PR TITLE
Fix #5683: False positive ``used-before-assignment`` in loop `else` where the only non-break exit is through `except`

### LINKED ISSUE #5683: False positive for `used-before-assignment` in `else` clause when the only non-`break` exit from a loop is an except block
### Bug description

```python
def do_something():
    pass

for _ in range(3):
    try:
        do_something()  # not an exit branch because of `else` below (will only exit here with an exception)
    except ZeroDivisionError as e:
        error = e  # < exit branch where error is defined
    else:
        break  # < exit condition where error is *not* defined, but will skip `raise error` below
    print("retrying...")
else:
    raise error
```

### Configuration

_No response_

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:13:10: E0601: Using variable 'error' before assignment (used-before-assignment)
```

### Expected behavior

No message. Identified by @martimlobao and discussed in detail https://github.com/PyCQA/pylint/pull/5402#issuecomment-1012791780.

Quoting from my [suggested](https://github.com/PyCQA/pylint/pull/5402#issuecomment-1013209351) solution:

So maybe this is the constraint to check for:

- if the use of `error` being tested is guarded by `else` in `for/else` (or `while/else`)
- if the definition of `error` in an `except` block is a descendant of the same `for/while`
- if that inner `else` breaks
- if there is no `continue` anywhere at all underneath the loop and before the break (other than inside another loop, but including nested conditions and try/excepts)

### Pylint version

```shell
pylint 2.13.0-dev0
astroid 2.9.2
Python 3.10.1 (v3.10.1:2cd268a3a9, Dec  6 2021, 14:28:59) [Clang 13.0.0 (clang-1300.0.29.3)]
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #5683

### CODE REVIEW

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:783]: Can you add the error this is ignoring? `mypy` has not yet implemented the feature, but we're trying not to do general type ignores.

Btw, isn't this returning the correct typing? DO we need to update `get_node_first_ancestor_of_type`?

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:803]: I'd prefer passing `other_node_try_except` as parameter. Can we do that?

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Thanks @jacobtylerwalls! That's quite a lot of code!

File: pylint/checkers/variables.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/variables.py:783]: I agree typing ``get_node_first_ancestor_of_type`` in a prior MR would probably make sense here. But it seems it's node_ancestors from astroid's ``NodeNG`` that should be typed ?
```python
def get_node_first_ancestor_of_type(
    node: nodes.NodeNG, ancestor_type: Union[Type[T_Node], Tuple[Type[T_Node]]]
) -> Optional[T_Node]:
    """Return the first parent node that is any of the provided types (or None)"""
    for ancestor in node.node_ancestors():
        if isinstance(ancestor, ancestor_type):
            return ancestor
    return None
```

File: pylint/checkers/variables.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/variables.py:825]: That was a whole journey 😄 !

File: tests/functional/u/use/used_before_assignment_issue4761.py
Comment (Pierre-Sassoulas) [inline tests/functional/u/use/used_before_assignment_issue4761.py:178]: I'm pretty sure we're going to still be surprised by the code that exists in the wild, but I'm also pretty confident we won't be surprised often. Those tests are thorough !

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, the typing issue is minor even if we might want to dig into it before merging (fix it then rebase on it to remove the typing that became uneeded?). Thank you @jacobtylerwalls , great addition to 2.13 !

File: pylint/checkers/variables.py
Comment (jacobtylerwalls) [reply pylint/checkers/variables.py]: Ah, it needs a `...`! I'll fix in another PR (and the similar function just below it)

Comment (DanielNoord) [reply pylint/checkers/variables.py]: Ah, that's a nasty oversight from me in the review. Good catch!

**Review by DanielNoord — APPROVED** [review]
Sorry! Should have changed my review to approved after the last comments were integrated!

**Comment by DanielNoord** [comment]
@jacobtylerwalls Just something I thought of, I see the confidence of `used-before-assignment` is `UNDEFINED`. I believe this message will likely always require some form of inference (right?). Should we update the confidence of the message to be `INFERENCE`?

**Comment by jacobtylerwalls** [comment]
What about a new level `CONTROL_FLOW`? It's not really node-inference dependent.

**Comment by DanielNoord** [comment]
I like that a lot!

**Comment by jacobtylerwalls** [comment]
Cool, I'll PR it unless you've already started 🏃🏻‍♂️

**Comment by DanielNoord** [comment]
No I haven't, would certainly welcome a PR 😄

**Comment by DanielNoord** [comment]
@jacobtylerwalls This seems to have landed in stale PR heaven 😄  If you could update the test output I will merge this asap!

**Review by DanielNoord — APPROVED** [review]

**Comment by DanielNoord** [comment]
Thanks!

**Pull request merged**
